### ISSUE TITLE
1.7.5 Commerce transaction reserved backorder throws InvalidTransactionStateException

### ISSUE THREAD

**Issue by sihendra** [stevebauman/inventory#35]
Hi @stevebauman,

When I tried reserving with backorder flag turned on, I got this exception:
```
Stevebauman\Inventory\Exceptions\InvalidTransactionStateException with message 'Transaction state: commerce-reserved cannot be changed to a: commerce-back-ordered state.'
```

These are the statements causing that error (I used php artisan tinker):
```
>>> $s = InventoryStock::find(1)
=> Stevebauman\Inventory\Models\InventoryStock {#707
     id: "1",
     created_at: "2015-07-23 07:05:16",
     updated_at: "2015-07-23 08:22:04",
     deleted_at: null,
     user_id: null,
     inventory_id: "1",
     location_id: "1",
     quantity: "20.00",
     aisle: null,
     row: null,
     bin: null,
   }
>>> $t = $s->newTransaction();
=> Stevebauman\Inventory\Models\InventoryTransaction {#691
     stock_id: "1",
     name: "",
   }
>>> $t->reserved(30, true);
Stevebauman\Inventory\Exceptions\InvalidTransactionStateException with message 'Transaction state: commerce-reserved cannot be changed to a: commerce-back-ordered state.'
```

Is there something I missed to enable backorder reservation?

Thanks,
Hendra

**Comment by stevebauman** [comment]
Hey, thanks for contacting me.

I definitely thought I had a test case for this situation. This looks like it's just a validation issue going from the reservation state. I'm going to troubleshoot and get back to you early tomorrow. Thanks!

**Comment by mpallante** [comment]
Hello, I locally solved the problem by adding

    $this::STATE_COMMERCE_RESERVED,

to the `validatePreviousState()` call in `backOrder()`. I don't get the exception anymore, but I don't understand what is the right workflow in the case of backorders.

If I call `reserved()` with a quantity greater than the stock quantity, the stock quantity is not reduced and the transaction goes to the `STATE_COMMERCE_BACK_ORDERED` state. I'm not sure this is what I expect: maybe I'd like to have part of the quantity removed from the stock, and part to be back ordered.

Then what's the right workflow? Let's assume I increase the stock quantity (either starting a new transaction with `ordered()`) or by using the `put()` method on the stock itself.

I assume I should call `fillBackOrder()`, but after I do that, the new state `STATE_COMMERCE_BACK_ORDER_FILLED` is a dead end: no other method accepts it as a starting state.

Otherwise, if I don't call `fillBackOrder()` but just run the `sold()` method, the transaction goes to the `STATE_SOLD` state, but no quantity is taken from the stock.

I'm really confused about this behavior and I can't understand if there is something wrong in my code.

Could you please explain better?

Thanks a lot!

**Comment by stevebauman** [comment]
@mpallante You're right.The issues you've described are the issues I've ran into while creating the transaction system. I need to rework the whole system in my opinion, as working with so many states gets confusing and difficult to use as you're not exactly sure of what to expect. Give me some time and hopefully I'll have `2.0` out with a redone transaction system.

However, you're always free to override the methods and improve upon it yourself.

**Comment by koapeadu** [comment]
Hi @stevebauman, how is v2.0 coming along? 
I really need some updates on this awesome package.
